src/Util/ExAuth.php

```diff
@@ -0,0 +1,304 @@
+						fwrite(STDOUT, pack('nn', 2, 0));
+						break;
+				}
+			} else {
+				$this->writeLog(LOG_NOTICE, 'invalid command string ' . $sData);
+				fwrite(STDOUT, pack('nn', 2, 0));
+			}
+		}
+	}
+
+	/**
+	 * @brief Check if the given username exists
+	 *
+	 * @param array $aCommand The command array
+	 */
+	private function isUser(array $aCommand)
+	{
+		$a = get_app();
+
+		// Check if there is a username
+		if (!isset($aCommand[1])) {
+			$this->writeLog(LOG_NOTICE, 'invalid isuser command, no username given');
+			fwrite(STDOUT, pack('nn', 2, 0));
+			return;
+		}
+
+		// Now we check if the given user is valid
+		$sUser = str_replace(array('%20', '(a)'), array(' ', '@'), $aCommand[1]);
+
+		// Does the hostname match? So we try directly
+		if ($a->get_hostname() == $aCommand[2]) {
+			$this->writeLog(LOG_INFO, 'internal user check for ' . $sUser . '@' . $aCommand[2]);
+			$found = dba::exists('user', ['nickname' => $sUser]);
+		} else {
+			$found = false;
+		}
+
+		// If the hostnames doesn't match or there is some failure, we try to check remotely
+		if (!$found) {
+			$found = $this->checkUser($aCommand[2], $aCommand[1], true);
+		}
+
+		if ($found) {
+			// The user is okay
+			$this->writeLog(LOG_NOTICE, 'valid user: ' . $sUser);
+			fwrite(STDOUT, pack('nn', 2, 1));
+		} else {
+			// The user isn't okay
+			$this->writeLog(LOG_WARNING, 'invalid user: ' . $sUser);
+			fwrite(STDOUT, pack('nn', 2, 0));
+		}
+	}
+
+	/**
+	 * @brief Check remote user existance via HTTP(S)
+	 *
+	 * @param string $host The hostname
+	 * @param string $user Username
+	 * @param boolean $ssl Should the check be done via SSL?
+	 *
+	 * @return boolean Was the user found?
+	 */
+	private function checkUser($host, $user, $ssl)
+	{
+		$this->writeLog(LOG_INFO, 'external user check for ' . $user . '@' . $host);
+
+		$url = ($ssl ? 'https' : 'http') . '://' . $host . '/noscrape/' . $user;
+
+		$data = z_fetch_url($url);
+
+		if (!is_array($data)) {
+			return false;
+		}
+
+		if ($data['return_code'] != '200') {
+			return false;
+		}
+
+		$json = @json_decode($data['body']);
+		if (!is_object($json)) {
+			return false;
+		}
+
+		return $json->nick == $user;
+	}
+
+	/**
+	 * @brief Authenticate the given user and password
+	 *
+	 * @param array $aCommand The command array
+	 */
+	private function auth(array $aCommand)
+	{
+		$a = get_app();
+
+		// check user authentication
+		if (sizeof($aCommand) != 4) {
+			$this->writeLog(LOG_NOTICE, 'invalid auth command, data missing');
+			fwrite(STDOUT, pack('nn', 2, 0));
+			return;
+		}
+
+		// We now check if the password match
+		$sUser = str_replace(array('%20', '(a)'), array(' ', '@'), $aCommand[1]);
+
+		// Does the hostname match? So we try directly
+		if ($a->get_hostname() == $aCommand[2]) {
+			$this->writeLog(LOG_INFO, 'internal auth for ' . $sUser . '@' . $aCommand[2]);
+
+			$aUser = dba::select('user', ['uid', 'password'], ['nickname' => $sUser], ['limit' => 1]);

```

Comment:
This does seem shorter if you really don't need to pass `array('uid', 'password')` etc.. in these

Reply:
What do you mean?

Reply:
I mean if you look at the dba calls in my PR they are like this:
`dba:select('contact', array('id'), array('uid' => $id, 'self' => false), array('limit' => 1));`
the passing of array calls instead of where you are enclosing in brackets ['uid', 'password'], I'm saying makes the line shorter if indeed the array() calls aren't required.

Reply:
This variant is new to me.

Reply:
It exists since PHP 5.4 released in March 2012. 😅